accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,152 @@
+
+In the context of trimming applications, there are two main concerns:
+
+1. Any code that is used by the application must be preserved in the application.
+2. Any code that is not used by the application can be removed from the application.
+
+Note the difference in "must" and "can" between those two concerns. An application that works is preferred over a smaller application that doesn't. Therefore it is critical that any necessary code must be preserved in the application.
+
+## Scenarios and User Experience
+
+Developers can run the ILLinker on their .NET application and be confident that the resulting application works correctly, while meeting size goals.
+
+The main application types that will be the initial focus are Xamarin and Blazor applications. While ensuring small application sizes in other form factors is important, we will focus our efforts on the application types in which size is critical.
+
+### Goals
+
+1. Ensure all required code is preserved by the linker.
+2. Allow Xamarin and Blazor applications to use the .NET Libraries and be as small as possible.
+
+### Non-Goals
+
+1. Optimizing application size in other form factors.
+
+## Design
+
+There are a number of work items the .NET Libraries Team needs to undertake in .NET 5. In general they fall into the two buckets outlined above:
+
+1. Linker-safety
+2. Smaller Size
+
+### Linker-safety
+
+Ensuring required code is preserved by the linker is not easy. .NET has a very rich Reflection capability that enables dynamic invocation of code. There are many cases where a static analysis tool (like the ILLinker) is not able to know which code is required and which isn't. For example, a .NET application could read a Type name from a file and create a new instance of that Type through Reflection. When the linker analyzes the application statically, it doesn't see any hard references to that Type and removes it. But when the application is executed, an error occurs because the Type is no longer available in the application.
+
+There are features in the .NET Libraries that will be very expensive, and in some cases virtually impossible, to make 100% linker-safe. A few examples are:
+- `System.ComponentModel.Composition`, `System.Composition`, and other dependency injection style libraries
+    - Contains features such as loading "parts" from a directory
+- `Microsoft.CSharp` and `System.Linq.Expressions`
+    - Use Reflection extensively to implement `dynamic` and expression trees
+- `System.Resources.ResourceManager`
+    - Resource files can have ResourceReader Type names embedded in them, which `ResourceManager` creates dynamically.
+
```

Comment:
A place it comes into play is if you are emitting a call to a method/property that is trimmed away. The linker may think that method is not needed since it isn't called statically.

Reply:
> emitting a call to a method/property that is trimmed away

Sure, but to use the emit APIs to do so, you need to use reflection to get the relevant members so that you can pass in the MemberInfo to invoke, so it would seem to be the same reflection case.

Maybe I'm splitting hairs. Doesn't really matter.

Reply:
I think you are underestimating the SRE scope. For example, you can use SRE with tokens which you can get from handles, signatures or even stream of bytes. You can also generate new assemblies (we don't have a good concept for that in linker).

Reply:
Saving assemblies isn't supported in core. And the vast majority use case doesn't use those niche features you describe. I would be surprised if anyone using the linker with .NET 5 breaks because of them, but sure, it's possible. I would not be surprised if someone breaks because they emit a call via a MethodInfo to something that gets removed, but that's about the MethodInfo rather than the emit.

Reply:
(For the purposes of this doc, of course, it doesn't matter and as fine to call out if you think it's important.  I'm just questioning it from a prioritization perspective.)

Reply:
>  you can use SRE with tokens which you can get from handles, signatures or even stream of bytes.

These sound like special instances of dynamic loading (of assemblies that the linker did not know about). It is same category as `Assembly.Load(byte[])` or `Assembly.LoadFile`. Dynamic loading is not compatible with linker by design. Maybe mention dynamic loading as an example of feature that won't work with linker, ever?

Reply:
> Maybe mention dynamic loading as an example of feature that won't work with linker, ever?

Added.